# CS-UY 3943 - Introduction of Data Science
## Week 2: Using the Command-line

*Please edit the cell below to add your name, NetID, assigned group, and group members for this assignment.* **Complete this section before proceeding to the assignment questions.**

Name: Lenny Remache
NetID: ljr381
Group (Letter): G 
Group Member 1: Jeremy 
Group Member 2: Ricky
***

#### Part 1 - Processing Loan Data
A file named **loans_data.csv** is included in your repository. This file consists of records that pertain to loan records in a local bank. There are 15 comma separated columns in this order: 

`CustNum`, `Amount.Requested`,	`Amount.Funded.By.Investors`,	`Interest.Rate`,	`Loan.Length`,	`Loan.Purpose`,	`Debt.To.Income.Ratio`,	`State`,	`Home.Ownership`,	`Monthly.Income`,	`FICO.Range`,	`Open.CREDIT.Lines`,	`Revolving.CREDIT.Balance`,	`Inquiries.in.the.Last.6.Months`, and	`Employment.Length`. 

These fields contain data of type `int`, `float`, and `string`, and you are also provided a file named **loans_data_columns.csv** in the repository containing all the column names for easy reference.  Answer the following questions using Linux/Unix bash commands. All questions can be answered in one line (sometimes, with pipes)! Some questions will have many possible solutions. In jupter notebooks, you must prefix all shell commands with an exclamation point, i.e. `!command argument(s)` in order to execute the commands. As an example, to output the contents of a file named **file1.txt** followed by the contents of a file named **file2.txt**, you could execute the following command in the shell:

`cat file1.txt file2.txt`

In order to replicate this command in Jupyter Notebook, execute the following in a code cell:

`!cat file1.txt file2.txt`

1\. How many records (lines) are in **loans_data.csv**?

In [4]:
!wc -l loans_data.csv

    2501 loans_data.csv


<p>
<span style="color:blue">
2501</span>
</p>

2\. How many unique `State` (the 8th field) values are in this file? (hint: consider the 'cut' command and use the pipe operator '|')

In [12]:
!cut -d , -f 8 loans_data.csv | sort | uniq | wc -l

      47


<p>
<span style="color:blue">
47</span>
</p>

3\. Rank the reasons (`Loan.Purpose` (the 6th field)) loans were requested in descending order including the count of each category. (HINT: consider the `cut`, `uniq` and `sort` commands and the pipe (`|`) operator).

In [9]:
!cut -d , -f 6 loans_data.csv | sort | uniq -c | sort -r | head -n 14

1307 "debt_consolidation"
 444 "credit_card"
 201 "other"
 152 "home_improvement"
 101 "major_purchase"
  87 "small_business"
  50 "car"
  39 "wedding"
  30 "medical"
  29 "moving"
  21 "vacation"
  20 "house"
  15 "educational"
   4 "renewable_energy"


4\. List all records which have `FICO.Range` (the 11th field) from 815-819. (hint: This can be done using `grep`. Use `man` for a list of the command-line arguments available for `grep`.)

In [12]:
!grep 815-819 loans_data.csv

"64884",9000,9000,"6.03%","36 months","vacation","5.58%","NJ","MORTGAGE",9583.33,"815-819",11,675,0,"n/a"
"55501",8000,8000,"6.03%","36 months","debt_consolidation","4.51%","OR","MORTGAGE",3500,"815-819",9,6737,0,"10+ years"
"93374",16500,16500,"6.03%","36 months","debt_consolidation","22.65%","CA","MORTGAGE",5416.67,"815-819",17,14835,0,"10+ years"
"90568",4800,4800,"6.62%","36 months","car","10.42%","TX","MORTGAGE",7291.67,"815-819",14,0,0,"< 1 year"
"80302",16800,16800,"7.90%","60 months","debt_consolidation","3.34%","FL","MORTGAGE",10666.67,"815-819",7,4757,0,"10+ years"
"5906",12800,12787.71,"8.94%","36 months","debt_consolidation","0.18%","AZ","MORTGAGE",2833.33,"815-819",7,306,0,"4 years"


#### Part 2 - Combining Data Files
A second file is present in your repository, within the `music` directory. This directory consists of 4 files: `tracks1.csv`, `tracks2.csv`, `tracks3.csv`, and `tracks4.csv`. Each file contains data about musical tracks. While all of the invidual the files contain the same columns, 

`track_id`, `bit_rate`, `comments`, `composer`, `date_created`, `date_recorded`, `duration`, `favorites`, `genre_top`, `genres`, `genres_all`, `information`, `interest`, `language_code`, `license`, `listens`, `lyricist`, `number`, `publisher`, `tags`, and `title`

it is more convinent to process the data as a single file. Combining the files into a single file can be accomplished using command line tools.


1\. Using `ls`, `wc`, and a pipe operator (`|`), write and execute a shell command that displays the number of `.csv` files contained in the `music` directory.

In [2]:
!ls ./music/*.csv | wc -l

       4


2\. Join the files together into a new file named **all_tracks.csv**. Perform the concatenation such that only the header from the first file (**tracks1.csv**) is added to the new file. For all other files, all lines **excluding** the header will be added to the new file. 

HINT: the operator `>>` can be used for redirection. This operator writes to the the file named on the right side of the operator such that data from the command on the left is appended to the file on the right. For example,

`echo "Inserting a line" >> output.txt`

will add the text `Inserting a line` to the end of the file **output.txt**. Whereas, 

`echo "Inserting a line" > output.txt`

will overwrite any previous data in **output.txt** and replace that data with the text `Inserting a line`.

1. Append the header (the first line) from **tracks1.csv** to **all_tracks.csv**.

In [7]:
!head -n 1 ./music/tracks1.csv >> all_tracks.csv

2. Write a `for` loop that utilizes shell commands to append all lines (excluding the header line) from the input "tracks" files (**tracks1.csv**, **tracks2.csv**, **tracks3.csv**, **tracks4.csv**). While in this particular case, we know that there are 4 tracks files, write your `for` loop so that it would work regardless of the number of tracks files available to append to **all_tracks.csv**.

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;For Windows users, use the following syntax to execute your `for` loop:

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`!bash -c "LOOP"`

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;where `LOOP` is replaced with the the `for` loop to execute.

In [4]:
!for file in ./music/*.csv; do tail -n+2 $file >> all_tracks.csv; done

3\. Confirm that **all_tracks.csv** contains all of the lines from the input files. Do this by executing a command showing the number of lines in all of the input files as well as the number of lines in **all_tracks.csv**. (The count of lines in **all_tracks.csv** will be one greater than the sum of the input files due to a newline character at the end of the files).

In [8]:
!wc -l ./music/*.csv  all_tracks.csv

   28851 ./music/tracks1.csv
   27602 ./music/tracks2.csv
   26977 ./music/tracks3.csv
   32344 ./music/tracks4.csv
  115773 all_tracks.csv
  231547 total


4\. Write and execute a shell command that will remove all of the files that were concatenated together as **all_tracks.csv**.

In [13]:
!rm ./music/*.csv

5\. From Terminal/Anaconda Prompt, add **all_tracks.csv** to your repository. Commit and push the changes to this Jupyter notebook and the **music** directory (which should now only contain a single file) to your repository.

In [16]:
!head ./music/*.csv

zsh:1: no matches found: ./music/*.csv
